### PR TITLE
[US159126]Updating to GHA, using node v20 and moving from node-sass to sass

### PR DESCRIPTION
**Pull request by venkata-ramana-d2l**
### Rally Link
https://rally1.rallydev.com/#/668026692137d/teamboard?detail=%2Fuserstory%2F729272559259

### Description
1. Using GHA
2. Updating to node v20
3. Moving from node-sass to sass
4. Using Chrome headless instead of PhantomJS

build sample

image

test

image

### CODE REVIEW

File: karma.conf.js [hunk 1]
```diff
@@ -1,5 +1,9 @@
 var isDev = process.env.NODE_ENV === 'development';
 
+const puppeteer = require('puppeteer');
+process.env.CHROME_BIN = puppeteer.executablePath();
+
+
```
Comment (prajapac) [inline karma.conf.js:6]: nit: extra newline

File: src/plugins/pdf/native/nativeViewer.js [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 var React = require('react'),
 	GenericViewer = require('../../generic/viewer.js'),
-	NativeViewer = require('react-valence-ui-iframe');
+	ImportedNativeViewer = require('react-valence-ui-iframe');
```
Comment (prajapac) [inline src/plugins/pdf/native/nativeViewer.js:5]: Why is this being renamed?

File: src/plugins/pdf/native/nativeViewer.js [hunk 3]
```diff
@@ -28,7 +28,7 @@ var NativeViewer = React.createClass({
 	},
 	render: function() {
 		return <div className="vui-fileviewer-pdf-native">
-			<NativeViewer {...props} />
+			<ImportedNativeViewer {...this.props} />
```
Comment (prajapac) [inline src/plugins/pdf/native/nativeViewer.js:31]: With the rename, why do we now need `this.`?

File: src/plugins/pdf/native/__tests__/supportsNativePdfTests.js [hunk 4]
```diff
@@ -6,12 +6,11 @@ describe('PDF Native Support', function() {
...
 		window.ActiveXObject = undefined;
-		navigator.mimeTypes = {};
 	});
 
-	it('should return false if no mimeType entry', function() {
+	it('should return true if there is mimeType entry', function() {
 		var result = supportsNativePdf();
-		expect(result).toBeFalsy();
+		expect(result).toBeTrue();
 	});
```
Comment (prajapac) [inline src/plugins/pdf/native/__tests__/supportsNativePdfTests.js:14]: Why did we get rid of `navigator.mimeTypes` and invert the logic of the test? 🤔

File: .github/workflows/ci.yml
Comment (prajapac) [inline .github/workflows/ci.yml:1]: The `travis.yml` file also ran `npm run report-coverage`, are we able to add that? What output do you get when you run it?

File: .github/workflows/release.yml
Comment (prajapac) [inline .github/workflows/release.yml:1]: Same comment as for the `ci.yml` file

File: karma.conf.js [hunk 5]
```diff
@@ -13,21 +17,21 @@ module.exports = function(config) {
 			dir : 'coverage',
 			reporters: [
 				{ type: 'text' },
-				{ type: 'lcov' }
+				{ type: 'html' }
```
Comment (prajapac) [inline karma.conf.js:20]: Are you able to explain the change here? 🤔

File: karma.conf.js [hunk 6]
```diff
@@ -13,21 +17,21 @@ module.exports = function(config) {
...
 			]
 		},
 		browserify: {
 			debug: isDev,
 			extensions: ['.js'],
-			transform: isDev ? [babelify] : [babelify, istanbul]
+			transform: isDev ? [babelify] : [babelify, istanbul],
+			plugin: ['esmify']
```
Comment (prajapac) [inline karma.conf.js:27]: What does this added plugin do?

File: karma.conf.js [hunk 7]
```diff
@@ -13,21 +17,21 @@ module.exports = function(config) {
...
 		},
 		// There's a pre-test stall on Travis builds that can cause Karma to
 		//  fail with the default timeout of 10s.  I'm not sure what's causing
 		//  the stall, but this mitigates it for now.
 		browserNoActivityTimeout: 30000,
-		browsers: ['PhantomJS'],
+		browsers: ['ChromeHeadless'],
 		files: [
-			'node_modules/phantomjs-polyfill/bind-polyfill.js',
```
Comment (prajapac) [inline karma.conf.js:30]: We don't need this file since we are deprecating use of `phantomjs`, but do we need an alternate `pollyfill`? Do you happen to know what this file did/why it was needed?

**Review by prajapac — APPROVED** [review]

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        uses: Brightspace/third-party-actions@actions/setup-node
+        with:
+            node-version-file: .nvmrc
+      - name: Semantic Release
+        uses: BrightspaceUI/actions/semantic-release@master
+        with:
+          DEFAULT_BRANCH: master
+          GITHUB_TOKEN: ${{ secrets.D2L_GITHUB_TOKEN }}
+          NPM: true
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Comment (prajapac) [inline .github/workflows/release.yml:26]: Do we have these tokens set up?

Comment (venkata-ramana-d2l) [reply .github/workflows/release.yml:26]: nope, that was actually my first question to you on slack

https://d2l.slack.com/archives/C05VARVDVPC/p1700139742597769

File: karma.conf.js (hunk 7 above)
Comment (venkata-ramana-d2l) [reply karma.conf.js:30]: its actually a polyfill for bind function as phantomjs did not have it - https://github.com/tom-james-watson/phantomjs-polyfill

as we are using chrome headless, it does not have this problem. So, we dont need a new polyfill.

File: karma.conf.js (hunk 6 above)
Comment (venkata-ramana-d2l) [reply karma.conf.js:27]: converts import/export statements which browserify does not understand.

https://stackoverflow.com/a/55176597

File: src/plugins/pdf/native/nativeViewer.js (hunk 2 above)
Comment (venkata-ramana-d2l) [reply src/plugins/pdf/native/nativeViewer.js:5]: because, the component is NativeViewer and we are using NativeViewer component inside it.

File: src/plugins/pdf/native/nativeViewer.js (hunk 3 above)
Comment (venkata-ramana-d2l) [reply src/plugins/pdf/native/nativeViewer.js:31]: yea.. because props is a property of the class and it has to be accessed via `this`

File: karma.conf.js (hunk 1 above)
Comment (venkata-ramana-d2l) [reply karma.conf.js:6]: done

File: karma.conf.js (hunk 5 above)
Comment (venkata-ramana-d2l) [reply karma.conf.js:20]: this was accidental, removed it.

File: src/plugins/pdf/native/__tests__/supportsNativePdfTests.js (hunk 4 above)
Comment (venkata-ramana-d2l) [reply src/plugins/pdf/native/__tests__/supportsNativePdfTests.js:14]: below is the error with having navigator.mimeTypes

image

Comment (venkata-ramana-d2l) [reply src/plugins/pdf/native/__tests__/supportsNativePdfTests.js]: And I reverted the test case as I realised the same is being checked in the test case that follows this one.

File: .github/workflows/ci.yml
Comment (venkata-ramana-d2l) [reply .github/workflows/ci.yml:1]: "report-coverage": "lcov-result-merger 'coverage/**/lcov.info' | ./node_modules/coveralls/bin/coveralls.js",

This script works with travis CI. Moreover, we dont need lcov-result-merger as there is only one lcov.info file for this repo. So, removed this script as well as lcov-result-merger.

And, added Report coverage task to ci.yml file that publishes the result to coveralls.io - https://coveralls.io/jobs/132145487

File: .github/workflows/release.yml
Comment (venkata-ramana-d2l) [reply .github/workflows/release.yml:1]: done

File: package.json
Comment (venkata-ramana-d2l) [inline package.json:1]: @prajapac should we change the version for the changes done in this PR? currently, its 1.1.0. Maybe make it 2.0.0?

File: karma.conf.js (hunk 7 above)
Comment (prajapac) [reply karma.conf.js:30]: Ahh ok, that makes sense!

File: package.json
Comment (prajapac) [reply package.json:1]: I'd say that makes sense, we are definitely doing some major rework! Let's do a major version bump to 2.0.0!

File: src/plugins/pdf/native/nativeViewer.js (hunk 2 above)
Comment (prajapac) [reply src/plugins/pdf/native/nativeViewer.js:5]: Ahh yes, I recall that linter complaint from a long time ago, but I didn't make any changes since I wasn't sure what's what, and it was hard to test this repo prior to the kind of changes that are being made as part of this PR.

File: src/plugins/pdf/native/nativeViewer.js (hunk 3 above)
Comment (prajapac) [reply src/plugins/pdf/native/nativeViewer.js:31]: I'm probably missing something here, I understand that `props` is  a property of the class this code is in, but it wasn't necessary before, so how did it work then? Was it not necessary? We should leave your change in for sure, I'm just trying to understand is all.🤔

File: karma.conf.js (hunk 6 above)
Comment (prajapac) [reply karma.conf.js:26]: What the plugin does makes sense to me, but similar to a previous comment, how come we didn't need it before and need it now? Again, just trying to understand the changes. Thanks!

File: src/plugins/pdf/native/__tests__/supportsNativePdfTests.js [hunk 9]
```diff
@@ -6,12 +6,12 @@ describe('PDF Native Support', function() {
...
 		window.ActiveXObject = undefined;
-		navigator.mimeTypes = {};
 	});
 
 	it('should return false if no mimeType entry', function() {
+		navigator.mimeTypes['application/pdf'] = undefined;
 		var result = supportsNativePdf();
-		expect(result).toBeFalsy();
+		expect(result).toBeFalse();
 	});
```
Comment (prajapac) [inline src/plugins/pdf/native/__tests__/supportsNativePdfTests.js:15]: The other test cases seem to have the old syntax still, i.e. `expect(result).toBeFalsy();`, any particular reason this one is different, i.e. `expect(result).toBeFale();`?

File: src/plugins/pdf/native/__tests__/supportsNativePdfTests.js [hunk 10]
```diff
@@ -6,12 +6,12 @@ describe('PDF Native Support', function() {
 
 	beforeEach(function() {
 		window.ActiveXObject = undefined;
-		navigator.mimeTypes = {};
 	});
 
 	it('should return false if no mimeType entry', function() {
+		navigator.mimeTypes['application/pdf'] = undefined;
```
Comment (prajapac) [inline src/plugins/pdf/native/__tests__/supportsNativePdfTests.js:12]: Why was this change necessary, i.e. not `navigator.mimeTypes` not being set before each function run, and then being set explicitly to `undefined` for the `application/pdf` mime type? 🤔

File: src/plugins/pdf/native/__tests__/supportsNativePdfTests.js (hunk 9 above)
Comment (venkata-ramana-d2l) [reply src/plugins/pdf/native/__tests__/supportsNativePdfTests.js]: it does not make a difference in this case. Both `toBeFalsy` and `toBeFalse` can be used as the method that this test case is testing returns boolean. However, changed it back to `toBeFalsy` to leave it untouched.

File: karma.conf.js (hunk 6 above)
Comment (venkata-ramana-d2l) [reply karma.conf.js:26]: image


So, the issue was coming with lit-element package which is a part of react-valence-ui-iframe package which in turn is a dependency of this repo. I see that lit-element was moved to dependencies from dev dependencies in react-valence-ui-iframe package [here](https://github.com/Brightspace/react-valence-ui-iframe/releases/tag/v3.0.3). So, am assuming this to be the cause for this issue.

Do you know if this repo's build was tested when the above change was made in react-valence-ui-iframe package?

File: src/plugins/pdf/native/nativeViewer.js (hunk 3 above)
Comment (venkata-ramana-d2l) [reply src/plugins/pdf/native/nativeViewer.js:31]: are you sure that it was working before? The reason I added `this.` was, `lint` cmd was throwing the below issue. It also makes sense to add it because otherwise props is not declared/defined anywhere else in the component.

image

File: .github/workflows/release.yml (hunk 8 above)
Comment (venkata-ramana-d2l) [reply .github/workflows/release.yml:30]: So, as we discussed [here](https://d2l.slack.com/archives/C05VARVDVPC/p1697049021304529), I want to try first by not publishing this repo to codeartifact. Only if it makes the npm i slow for either activity-viewer repo or lessons repo, I want to publish this to codeartifact.

File: package.json [hunk 11]
```diff
@@ -1,16 +1,15 @@
...
+  "version": "2.0.0",
   "description": "React component for viewing files",
   "main": "src/fileViewer.js",
   "scripts": {
     "prebuild-sample": "rimraf ./sample/dist && mkdirp ./sample/dist",
     "build-sample": "npm run build-sample:js && npm run build-sample:css",
-    "build-sample:js": "browserify -s app ./sample/app.js > ./sample/dist/app.js",
-    "build-sample:css": "node-sass --importer ./node_modules/frau-sass-importer/ ./sample/app.scss ./sample/dist/app.css",
+    "build-sample:js": "browserify -s app ./sample/app.js -p esmify > ./sample/dist/app.js",
+    "build-sample:css": "sass -I node_modules -I ./ ./sample/app.scss  ./sample/dist/app.css",
```
Comment (prajapac) [inline package.json:10]: Since we are no longer using frau-sass-importer, should this line now be:
`"build-sample:css": "sass -I ./sample/app.scss -I ./sample/dist/app.css",`? 🤔

File: src/plugins/pdf/native/__tests__/nativeViewerTests.js [hunk 12]
```diff
@@ -9,6 +9,7 @@ describe('PDF Native Viewer', function() {
 
 	beforeEach(function() {
 		NativeViewer.__Rewire__('GenericViewer', 'div');
+		NativeViewer.__Rewire__('ImportedNativeViewer', 'div');
```
Comment (prajapac) [inline src/plugins/pdf/native/__tests__/nativeViewerTests.js:12]: One final question, what's up with the addition of this line here?

File: package.json (hunk 11 above)
Comment (venkata-ramana-d2l) [reply package.json:10]: `-I` is for loading scss paths that are used by the scss files.

by the above command, we are saying, we need some scss files from node_modules path and the root path of the project repo. 

This is because, certain css from bower components is being used here 
image

and a particular scss file is being used from react-valence-ui-iframe (which is from node_modules)

image

File: package.json [hunk 13]
```diff
@@ -1,16 +1,15 @@
 {
   "name": "react-vui-fileviewer",
-  "version": "1.1.0",
+  "version": "2.0.0",
```
[the remark this answers is not in the document]
Comment (venkata-ramana-d2l) [reply package.json:3]: re-generated package-lock.
